src/zkp/pifac.rs

```diff
@@ -156,64 +196,82 @@ impl Proof for PiFacProof {
+                .reconstruct(&self.p_masked, &self.masked_p_commitment_randomness);
+            let rhs = input.setup_params.scheme().combine(
+                &self.p_mask_commitment,
+                &self.p_commitment,
+                &e,
+            );
             lhs == rhs
         };
-        if !eq_check_1 {
+        if !masked_p_commitment_is_valid {
             error!("eq_check_1 failed");
             return Err(InternalError::ProtocolError);
         }
 
-        let eq_check_2 = {
-            let lhs = input.setup_params.scheme().reconstruct(&self.z2, &self.w2);
-            let rhs = input.setup_params.scheme().combine(&self.B, &self.Q, &e);
+        let masked_q_commitment_is_valid = {
+            let lhs = input
+                .setup_params
+                .scheme()
+                .reconstruct(&self.q_masked, &self.masked_q_commitment_randomness);
+            let rhs = input.setup_params.scheme().combine(
+                &self.q_mask_commitment,
+                &self.q_commitment,
+                &e,
+            );
             lhs == rhs
         };
-        if !eq_check_2 {
+        if !masked_q_commitment_is_valid {
             error!("eq_check_2 failed");
             return Err(InternalError::ProtocolError);
         }
 
-        let eq_check_3 = {
+        let modulus_links_provided_factors = {
             let R = input
                 .setup_params
                 .scheme()
-                .reconstruct(&input.N0, self.sigma.as_masked());
-            let lhs = input
+                .reconstruct(&input.modulus, self.sigma.as_masked());
+            let lhs = input.setup_params.scheme().reconstruct_with_commitment(
+                &self.q_commitment,
+                &self.p_masked,
+                &self.masked_p_link,
+            );
+            let rhs = input
                 .setup_params
                 .scheme()
-                .reconstruct_with_commitment(&self.Q, &self.z1, &self.v);
-            let rhs = input.setup_params.scheme().combine(&self.T, &R, &e);
+                .combine(&self.q_link_commitment, &R, &e);
             lhs == rhs
         };
-        if !eq_check_3 {
+        if !modulus_links_provided_factors {
             error!("eq_check_3 failed");
             return Err(InternalError::ProtocolError);
         }
 
-        let sqrt_N0 = sqrt(&input.N0);
+        let sqrt_N0 = sqrt(&input.modulus);
         // 2^{ELL + EPSILON}
         let two_ell_eps = BigNumber::one() << (ELL + EPSILON);
         // 2^{ELL + EPSILON} * sqrt(N_0)
         let z_bound = &sqrt_N0 * &two_ell_eps;
-        if self.z1 < -z_bound.clone() || self.z1 > z_bound {
+        if self.p_masked < -z_bound.clone() || self.p_masked > z_bound {
             error!("self.z1 > z_bound check failed");
             return Err(InternalError::ProtocolError);
         }
-        if self.z2 < -z_bound.clone() || self.z2 > z_bound {
+        if self.q_masked < -z_bound.clone() || self.q_masked > z_bound {
```

Comment:
Oh, now that I look at that function more closely, it checks that the parameter is in the range `± 2^n`. The check we need in this proof doing here is that the parameter is in the range `±√modulus · 2 ^ (ℓ+ε)`. So that function actually doesn't work as-is.

Please either write a separate utility function that tests ranges of the form we need here or revert back to doing it inline here.

Reply:
The change I recommended makes this check incorrect, this needs to be changed.